research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -199,9 +188,71 @@ def get_model(self, config: Config) -> nn.Module:
     def get_criterion(self, config: Config) -> _Loss:
         return Module2LossWrapper(self.nnunet_trainer.loss)
 
+    def get_lr_scheduler(self, config: Config) -> _LRScheduler:
+        """
+        Creates an LR Scheduler similar to the nnunet default except we set max steps
+        to the total number of steps and update every step. Initial and final LR are
+        the same as nnunet, difference is nnunet sets max steps to num 'epochs', but
+        they define an 'epoch' as exactly 250 steps. Therefore they update every 250
+        steps. Override this method to set your own LR scheduler.
+
+        Args:
+            config (Config): The server config. This method will look for the
+        Returns:
+            _LRScheduler: The default nnunet LR Scheduler for nnunetv2 2.5.1
+        """
+        if not isinstance(self.nnunet_trainer.lr_scheduler, PolyLRScheduler):
+            log(
+                WARN,
+                (
+                    "Nnunet seems to have changed their default LR scheduler to "
+                    f"type: {type(self.nnunet_trainer.lr_scheduler)}. "
+                    "Using PolyLRScheduler instead. Override or update the "
+                    "get_lr_scheduler method of nnUNetClient to change this"
+                ),
+            )
+
+        # Determine total number of steps throughout all FL rounds
+        local_epochs, local_steps, _, _ = self.process_config(config)
+        if local_steps is not None:
+            self.steps_per_round = local_steps
+            self.total_steps = int(config["n_server_rounds"]) * self.steps_per_round
+        elif local_epochs is not None:
+            self.steps_per_round = local_epochs * len(self.train_loader)
+            self.total_steps = int(config["n_server_rounds"]) * self.steps_per_round
+
+        # Create and return LR Scheduler. This is nnunet default for version 2.5.1
+        return PolyLRScheduler(
+            self.optimizers["global"], initial_lr=self.nnunet_trainer.initial_lr, max_steps=self.total_steps
+        )
+
     def get_optimizer(self, config: Config) -> Optimizer:
         return self.nnunet_trainer.optimizer
 
+    def get_max_batch_voxels(self, source_plans: Dict, max_factor: float = 0.05) -> float:
+        """
+        Determines the maximum number of voxels that should be in a batch
+
+        Args:
+            source_plans (Dict): The nnunet plans dict that is being modified
+            max_factor (float): The fraction of the approx total number of voxels to
+                use as the maximum. Defaults to 0.05 corresponding to 5% as this is the nnunet default in 2.5.1
+
+        Returns:
+            float: Maximum number of voxels that can be in a batch
+        """
+        # Need to determine input dimensionality
```

Comment:
It seems as though this if statement is doing the same thing regardless of NnUNetConfig._3D_FULLRES.value

Reply:
Good catch